lib/Dialect/ESI/runtime/cosim/lib/RpcServer.cpp

```diff
@@ -0,0 +1,355 @@
+  auto reactor = context->DefaultReactor();
+  reactor->Finish(Status::OK);
+  return reactor;
+}
+
+namespace {
+/// When a client connects to a read port (on its end, a write port on this
+/// end), construct one of these to poll the corresponding write port on this
+/// side and forward the messages.
+class RpcServerWriteReactor : public ServerWriteReactor<esi::cosim::Message> {
+public:
+  RpcServerWriteReactor(RpcServerWritePort *writePort)
+      : writePort(writePort), sentSuccessfully(SendStatus::UnknownStatus),
+        shutdown(false) {
+    myThread = std::thread(&RpcServerWriteReactor::threadLoop, this);
+  }
+  ~RpcServerWriteReactor() {
+    shutdown = true;
+    myThread.join();
+  }
+
+  void OnDone() override { delete this; }
+  void OnWriteDone(bool ok) override {
+    std::scoped_lock<std::mutex> lock(sentMutex);
+    sentSuccessfully = ok ? SendStatus::Success : SendStatus::Failure;
+    sentSuccessfullyCV.notify_one();
+  }
+  void OnCancel() override {
+    std::scoped_lock<std::mutex> lock(sentMutex);
+    sentSuccessfully = SendStatus::Disconnect;
+    sentSuccessfullyCV.notify_one();
+  }
+
+private:
+  /// The polling loop.
+  void threadLoop();
+  /// The polling thread.
+  std::thread myThread;
+
+  /// Assoicated write port on this side. (Read port on the client side.)
+  RpcServerWritePort *writePort;
+
+  /// Mutex to protect the sentSuccessfully flag.
+  std::mutex sentMutex;
+  enum SendStatus { UnknownStatus, Success, Failure, Disconnect };
+  volatile SendStatus sentSuccessfully;
+  std::condition_variable sentSuccessfullyCV;
+
+  volatile bool shutdown;
+};
+
+} // namespace
+
+void RpcServerWriteReactor::threadLoop() {
+  while (!shutdown && sentSuccessfully != SendStatus::Disconnect) {
+    // TODO: adapt this to a new notification mechanism which is forthcoming.
+    if (writePort->writeQueue.empty())
+      std::this_thread::sleep_for(std::chrono::microseconds(100));
+
+    // This lambda will get called with the message at the front of the queue.
+    // If the send is successful, return true to pop it. We don't know, however,
+    // if the message was sent successfully in this thread. It's only when the
+    // `OnWriteDone` method is called by gRPC that we know. Use locking and
+    // condition variables to orchestrate this confirmation.
+    writePort->writeQueue.pop([this](const MessageData &data) -> bool {
+      esi::cosim::Message msg;
+      msg.set_data(reinterpret_cast<const char *>(data.getBytes()),
+                   data.getSize());
+
+      // Get a lock, reset the flag, start sending the message, and wait for the
+      // write to complete or fail. Be mindful of the shutdown flag.
+      std::unique_lock<std::mutex> lock(sentMutex);
+      sentSuccessfully = SendStatus::UnknownStatus;
+      StartWrite(&msg);
+      while (!shutdown && sentSuccessfully == SendStatus::UnknownStatus)
+        sentSuccessfullyCV.wait_for(lock, std::chrono::milliseconds(10));
```

Comment:
Why poll? Seems like you have full control of when `sentSuccessfully` and `shutdown` are assigned; and as such, it should be possible to do the following, assuming the `shutdown` assignment on line 130 is placed behind the mutex and a cv wakeup:
```c++
sentSuccessfullyCV.wait(lock, [&](){return shutdown || sentSuccessfully != SendStatus::UnknownStatus});
```

Reply:
Actually, I don't think I need to lock it given that shutdown is an atomic assignment. Great suggestion.